components/service/glean/src/main/java/mozilla/components/service/glean/scheduler/PingScheduler.kt

```diff
@@ -0,0 +1,206 @@
+ * [Lifecycle.Event.ON_STOP] state, it will use [WorkManager] to schedule a [PingUploadWorker] task
+ * which is responsible for uploading the ping.
+ */
+internal class PingScheduler(
+    applicationContext: Context,
+    storageEngineManager: StorageEngineManager,
+    internal var pingMaker: PingMaker = PingMaker(storageEngineManager, applicationContext),
+    internal var pingStorageEngine: PingStorageEngine = PingStorageEngine(applicationContext),
+    internal var metricsPingScheduler: MetricsPingScheduler = MetricsPingScheduler(applicationContext)
+) : LifecycleObserver {
+    private val logger = Logger("glean/GleanLifecycleObserver")
+
+    // The application id detected by glean to be used as part of the submission
+    // endpoint.
+    internal var applicationId = sanitizeApplicationId(applicationContext.packageName)
+
+    companion object {
+        /**
+         * Helper function to check to see if a particular ping has been scheduled by the PingScheduler
+         *
+         * @return True if [PingUploadWorker] task found in [WorkManager], false otherwise
+         */
+        internal fun isWorkScheduled(): Boolean {
+            val instance = WorkManager.getInstance()
+            val statuses = instance.getWorkInfosByTag(PingUploadWorker.PING_WORKER_TAG)
+            return try {
+                var running = false
+                val workInfoList = statuses.get()
+                for (workInfo in workInfoList) {
+                    val state = workInfo.state
+                    running = (state === WorkInfo.State.RUNNING) or (state === WorkInfo.State.ENQUEUED)
+                }
+                running
+            } catch (e: ExecutionException) {
+                // Do nothing but return false
+                false
+            } catch (e: InterruptedException) {
+                // Do nothing but return false
+                false
+            }
+        }
+    }
+
+    init {
+        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
+    }
+
+    /**
+     * Handles necessary glean tasks upon entering the background state [Lifecycle.Event.ON_STOP]
+     */
+    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
+    fun onEnterBackground() {
+        // We're going to background, so store how much time we spent
+        // on foreground.
+        GleanBaseline.duration.stopAndSum()
+
+        if (!Glean.isInitialized()) {
+            logger.error("Glean not initialized, abort serialize and send ping")
+            return
+        }
+
+        // Do not send ping if upload is disabled
+        if (!Glean.getUploadEnabled()) {
+            logger.debug("Glean upload not enabled, abort serialize and send ping.")
+            return
+        }
+
+        // Schedule the baseline and event pings and save the result to determine whether or not
+        // we need to schedule the upload worker
+        var pingsToSchedule = serializePing(BaselinePing.STORE_NAME)
```

Comment:
nit: this variable is not telling us which pings to schedule, but whether or not *any* ping was assembled/is ready to be sent. Maybe `availableForSending` is a more suitable name?